### PR TITLE
Add support for nested cursor field, fixes #6016

### LINKED ISSUE #6016: Jira source normalization step error
OS Version / Instance: Ubuntu 20.04
Memory / Disk: 7.5Gb
Deployment: Docker
Airbyte Version: 0.29.17-alpha
Source name/version: Jira 0.2.10
Destination name/version: Postgresql 0.3.11
Step: Sync job
Description:

## Current Behavior

When basic normalization is enabled for the Jira source connector the job fails with the following error 
`io.airbyte.workers.WorkerException: Normalization process wasn't successful`

## Expected Behavior

After successful syncing, the normalization step should be successful.

## Logs
[airbyte_log.txt](https://github.com/airbytehq/airbyte/files/7153733/airbyte_log.txt)

## Steps to Reproduce
1. docker-compose up
2. Setup Source (Jira) and Destination (PostgreSQL)
3. Setup Connection using the source/destination created above, choose `issues` as stream and enable basic normalization

### PR DESCRIPTION
**Pull request by eliziario**
## What
fixes #6016 and adds support for defining a nested field as the cursor field

## How
generates the adequate field expressions when generating the scd.sql file during normalization

## Recommended reading order
1. `x.java`
2. `y.python`

## Pre-merge Checklist
Expand the relevant checklist and delete the others. 

<details><summary> <strong> New Connector </strong></summary>
<p>
   
#### Community member or Airbyter
   
- [ ] **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret` 
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated 
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] `docs/SUMMARY.md`
    - [ ] `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - [ ] `docs/integrations/README.md`
    - [ ] `airbyte-integrations/builds.md`
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [ ] Connector added to connector index like described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
   
#### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items. 
   
- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] Credentials added to Github CI. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci). 
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing. 
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
   
</p>
</details>

<details><summary> <strong> Updating a connector </strong></summary>
<p>
   
#### Community member or Airbyter
   
- [ ] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret` 
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated 
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [ ] Connector version bumped like described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
   
#### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items. 
   
- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] Credentials added to Github CI. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci). 
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing. 
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)

</p>
</details>

<details><summary> <strong> Connector Generator </strong> </summary>
<p>
   
- [ ] Issue acceptance criteria met
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [ ] If adding a new generator, add it to the [list of scaffold modules being tested](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connector-templates/generator/build.gradle#L41)
- [ ] The generator test modules (all connectors with `-scaffold` in their name) have been updated with the latest scaffold by running `./gradlew :airbyte-integrations:connector-templates:generator:testScaffoldTemplates` then checking in your changes
- [ ] Documentation which references the generator is updated as needed.
</p>
</details>

### CODE REVIEW

File: airbyte-integrations/connectors/source-jira/source_jira/streams.py [hunk 1]
```diff
@@ -203,20 +203,23 @@ def request_params(self, stream_state=None, **kwargs):
...
+        if stream_state.get(normalized_cursor_name):
+            issues_state = pendulum.parse(stream_state.get(normalized_cursor_name))
             issues_state_row = issues_state.strftime("%Y/%m/%d %H:%M")
             params["jql"] = f"updated > '{issues_state_row}'"
         return params
 
     def get_updated_state(self, current_stream_state: MutableMapping[str, Any], latest_record: Mapping[str, Any]) -> Mapping[str, Any]:
-        latest_record_date = pendulum.parse(latest_record.get("fields", {}).get(self.cursor_field))
+        normalized_cursor_name = ".".join(self.cursor_field)
+        print(latest_record.get("fields", {}))
```
Comment (keu) [inline airbyte-integrations/connectors/source-jira/source_jira/streams.py:215]:
```suggestion
```

File: airbyte-integrations/connectors/source-jira/source_jira/streams.py [hunk 2]
```diff
@@ -203,20 +203,23 @@ def request_params(self, stream_state=None, **kwargs):
...
+        normalized_cursor_name = ".".join(self.cursor_field)
+        if stream_state.get(normalized_cursor_name):
+            issues_state = pendulum.parse(stream_state.get(normalized_cursor_name))
             issues_state_row = issues_state.strftime("%Y/%m/%d %H:%M")
             params["jql"] = f"updated > '{issues_state_row}'"
         return params
 
     def get_updated_state(self, current_stream_state: MutableMapping[str, Any], latest_record: Mapping[str, Any]) -> Mapping[str, Any]:
-        latest_record_date = pendulum.parse(latest_record.get("fields", {}).get(self.cursor_field))
+        normalized_cursor_name = ".".join(self.cursor_field)
```
Comment (keu) [inline airbyte-integrations/connectors/source-jira/source_jira/streams.py:214]: I'm not sure I understand why do we need this `normalized` name here.

**Comment by sherifnada** [comment]
requesting review from @ChristopheDuong as the normalization master

File: airbyte-integrations/bases/base-normalization/normalization/transform_catalog/stream_processor.py [hunk 3]
```diff
@@ -604,7 +605,12 @@ def get_cursor_field(self, column_names: Dict[str, Tuple[str, str]], in_jinja: b
                 # using an airbyte generated column
                 cursor = self.cursor_field[0]
         else:
-            raise ValueError(f"Unsupported nested cursor field {'.'.join(self.cursor_field)} for stream {self.stream_name}")
+            # Dealing with nested fields
+            cursor = reduce(
+                lambda expression, next_item: f"{expression}->'{next_item}'",
```
Comment (ChristopheDuong) [inline airbyte-integrations/bases/base-normalization/normalization/transform_catalog/stream_processor.py:610]: not sure of this syntax with `->`... is this specific to a certain destination? which one?

File: airbyte-integrations/connectors/source-jira/source_jira/streams.py (hunk 2 above)
Comment (ChristopheDuong) [reply airbyte-integrations/connectors/source-jira/source_jira/streams.py:214]: same here

**Review by ChristopheDuong — CHANGES_REQUESTED** [review]
The parent issue that is linked to the PR seems to deal with an unknown `updated` column (is it a column not defined in the catalog returned by jira source but used as a cursor anyway?)

I don't see the relationship with nested cursor fields

maybe are you referring to this issue that was already closed? https://github.com/airbytehq/airbyte/issues/5088

**Comment by eliziario** [comment]
Hello @ChristopheDuong I am checking this again focusing in the catalog.

File: airbyte-integrations/bases/base-normalization/normalization/transform_catalog/stream_processor.py (hunk 3 above)
Comment (eliziario) [reply airbyte-integrations/bases/base-normalization/normalization/transform_catalog/stream_processor.py:610]: Yes, you're right, this is pg specific, it is the sintax to select the nested field in a json field. but anyway, your comments about the catalog coming from the source where enlightening, I am trying another approach.

**Comment by eliziario** [comment]
It looks like this is not the correct way to handle #6016. Closing this without merging

**Pull request closed without merging**